### ISSUE TITLE
tf.cumprod's gradient produces nans given zeros

### ISSUE THREAD

**Issue by girving** [tensorflow/tensorflow#3862]
Example:

```
tf.reset_default_graph()
var = tf.Variable([0.])
cumprod = tf.cumprod(var)
grad = tf.gradients(cumprod, var)
init = tf.initialize_all_variables()
with tf.Session() as sess:
  sess.run(init)
  print(sess.run(grad))

---> [array([ nan], dtype=float32)]
```

@ibab Do you know the right way to fix this?

**Comment by girving** [comment]
Cc @davidsaxton.

**Comment by ibab** [comment]
I've placed a comment on this in the gradient code here: [`tensorflow/python/ops/math_grad.py#L889`](https://github.com/tensorflow/tensorflow/blob/51fd9c024c4544ba1ef60862ec3f55b6e3ae79b1/tensorflow/python/ops/math_grad.py#L889)

I've spent some time trying to come up with a version that doesn't use a division by `x`, but didn't  come up with anything elegant.

**Comment by girving** [comment]
Hmm, that is pretty awkward:

    v = [x, y, z]
    cumprod = [x, xy, xyz]
    grad = [dx + y dy + yz dz, x dy + xz dz, xy dz]

**Comment by girving** [comment]
So in general you need all prefix products with each possible entry excluded.  Unfortunately if you write those as products of contiguous ranges you need quadratically many ranges.  Ug.

**Comment by khaotik** [comment]
A fast O(n) solution.
Assume the input vector `x` and the upstream gradient `g` has the form:
```
x = [x1 x2 x3 ... xn, 0, -, -, ...]
g = [g1 g2 g3 ... gn, gnp1, -, -, ...]
```
`-` Elements are irrelevant in this computation because the leftmost 0 cause `cumprod` become 0 after that.

```
[ dx1 ... dx <use unsafe method> ] + [ cumprod(x1,x2,x3...xn) * gnp1 ] + [0, 0, 0 ... 0]
```

To gain full speedup, the forward computation should record the index of first occurrence of `0`, and automatically `memset(0)` for the results after that. This optimization is unique to `cumprod` and can't be done to `cumsum`. The backward computation can just use that index to quickly do the job.

Sorry, not familiar with C++, so no PR yet.

**Comment by akosiorek** [comment]
The following workaround helps to avoid the NaN gradient issue (it lacks the `exclusive` flag but it's easy enough to implement):
```python
def cumprod(tensor, axis=0):
    transpose_permutation = None
    n_dim = len(tensor.get_shape())
    if n_dim > 0 and axis != 0:

        if axis < 0:
            axis = n_dim + axis

        transpose_permutation = np.arange(n_dim)
        transpose_permutation[-1], transpose_permutation[0] = 0, axis

    tensor = tf.transpose(tensor, transpose_permutation)

    def prod(acc, x):
        return acc * x

    prob = tf.scan(prod, tensor)
    tensor = tf.transpose(prob, transpose_permutation)
    return tensor
```

**Comment by gunan** [comment]
Automatically closing due to lack of recent activity. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!

It looks like there is a workaround posted.

**Comment by SamuelGabriel** [comment]
I submitted a fix to this issue in the PR #32714
